File: src/lib/types/paper.ts
export interface Paper {
  title: string;
  journal: string;
  year: string;
  authors: string;
  doi: string;
  openAccessUrl?: string; // Direct OA URL from OpenAlex/Unpaywall
}

export interface DownloadResult {
  success: boolean;
  source?: DownloadSource;
  filePath?: string;
  error?: string;
}

export type DownloadSource =
  | 'scihub'
  | 'annas-archive'
  | 'libgen'
  | 'unpaywall'
  | 'openalex-oa'
  | 'doi-direct';

export interface FailedDownload {
  paper: Paper;
  error: string;
  attemptedSources: DownloadSource[];
}

File: internal/types/paper.go
package types

type Paper struct {
	Title         string `json:"title"`
	Journal       string `json:"journal"`
	Year          string `json:"year"`
	Authors       string `json:"authors"`
	DOI           string `json:"doi"`
	OpenAccessURL string `json:"openAccessUrl,omitempty"`
	Abstract      string `json:"abstract,omitempty"`
}

type DownloadSource string

const (
	SourceScihub      DownloadSource = "scihub"
	SourceAnnasArch   DownloadSource = "annas-archive"
	SourceLibgen      DownloadSource = "libgen"
	SourceUnpaywall   DownloadSource = "unpaywall"
	SourceOpenAlexOA  DownloadSource = "openalex-oa"
	SourceDOIDirect   DownloadSource = "doi-direct"
)

type DownloadResult struct {
	Success  bool           `json:"success"`
	Source   DownloadSource `json:"source,omitempty"`
	FilePath string         `json:"filePath,omitempty"`
	Error    string         `json:"error,omitempty"`
}

type FailedDownload struct {
	Paper             Paper            `json:"paper"`
	Error             string           `json:"error"`
	AttemptedSources  []DownloadSource `json:"attemptedSources"`
}
